File: baekjoon/2293_동전_1.py
import sys
sys.stdin = open('2293.txt', 'r')

n, k = map(int, input().split())
nums = []

for i in range(n):
    num = int(input())
    nums.append(num)

former_result = [0] * (k + 1)
result = [0] * (k + 1)

former_result[0] = 1
result[0] = 1

coin_idx = 0

while coin_idx < n:
    for target in range(1, k + 1):
        if nums[coin_idx] > target:
            result[target] = former_result[target]
        else:
            result[target] = result[target - nums[coin_idx]] + former_result[target]
    former_result = result
    coin_idx += 1
    
print(result[k])
        
    

    

File: baekjoon/2293_동전_1.cpp
// 배열 초기화 하지 않으면 쓰레기값이 들어간다.
#include <iostream>

using namespace std;

int main()
{
    int n, k;
    int coins[101];

    cin >> n >> k;

    for (int i = 0; i < n; i++)
    {
        cin >> coins[i];
    }

    int result[10001] = {0};
    result[0] = 1;

    int coin_idx = 0;

    while (coin_idx < n)
    {
        for (int target = 1; target <= k; target++)
        {
            if (coins[coin_idx] <= target)
                result[target] += result[target - coins[coin_idx]];
        }
        coin_idx++;
    }

    cout << result[k] << endl;

    return 0;
}